src/Silex/Provider/FormServiceProvider.php

```diff
@@ -104,11 +107,19 @@ public function register(Application $app)
         });
 
         $app['form.csrf_provider'] = $app->share(function ($app) {
-            if (isset($app['session'])) {
-                return new SessionCsrfProvider($app['session'], $app['form.secret']);
-            }
+            if (!class_exists('Symfony\Component\Form\Extension\DataCollector\DataCollectorExtension')) {
+                // Symfony 2.3
+                if (isset($app['session'])) {
+                    return new SessionCsrfProvider($app['session'], $app['form.secret']);
+                }
 
-            return new DefaultCsrfProvider($app['form.secret']);
+                return new DefaultCsrfProvider($app['form.secret']);
+            } else {
+                // Symfony 2.4+
+                $storage = isset($app['session']) ? new SessionTokenStorage($app['session']) : new NativeSessionTokenStorage();
+
+                return new CsrfTokenManager(null, $storage);
+            }
```

Comment:
Technically, this is a BC break for people using the ``form.csrf_provider`` on their own, because it changes the class they get.
It might be better to register the CsrfTokenManager as a new service, and providing the BC adapter in ``form.csrf_provider`` (which would trigger a deprecation warning when they get it, but Silex itself would access the new one)